### ISSUE TITLE
Changing KeepFileOpen at runtime

### ISSUE THREAD

**Issue by 22u2** [NLog/NLog#1945]
I wanted to write some unit tests with xUnit and check that some strings exist in the Nlog target file but I have the issue saying that the File is locked but a process. This is due to KeepFileOpen is sets to true, we want performance. So my idea was to change this flag in the test setup (+ call to LogManager.ReconfigExistingLoggers() ) but it does not work as expected, I can't read the file with File.ReadAllText(myFile). In debug, I can that FileTarget is well updated, KeepFileOne = false.
So can we change the KeepFileOpen at runtime? or did i miss something?

**Type** :
- Question

**NLog version**: 4.4.2

**Platform**: .Net 4.5 

**Current NLog config** (xml or C#, if relevant)
```c#
var config = new LoggingConfiguration();

var myTarget= new FileTarget(myTargetFileName);
myTarget.Encoding = Encoding.UTF8;
myTarget.NetworkWrites = false;
myTarget.ConcurrentWrites = true;
myTarget.KeepFileOpen = true;

config.AddTarget(myTarget);

var myRule= new LoggingRule(myTargetFileName, LogLevel.Debug, myTarget);
config.LoggingRules.Add(myRule);

LogManager.Configuration = config;
```

Function to set the KeepFileOpen
```
FileTarget target = LogManager.Configuration.FindTargetByName(myTargetName) as FileTarget;
target.KeepFileOpen = false;
LogManager.ReconfigExistingLoggers()
```

**Comment by snakefoot** [comment]
> I wanted to write some unit tests with xUnit and check that some strings exist in the Nlog target file but I have the issue saying that the File is locked

Sounds like you are trying to open the file with exclusive access, not allowing other processes to write the file. Maybe try opening the file in non-exclusive mode?

Alternative then you can insert this line of code after the test-operation, but just before the test-verification:

`LogManager.Configuration = null;`

It will close all file-handles and then you can open the file in exclusive-mode.

**Comment by 22u2** [comment]
Thx for your solutions! But I still have issues.

> Sounds like you are trying to open the file with exclusive access, not allowing other processes to write the file. Maybe try opening the file in non-exclusive mode?

I tried with this:
```var stream = File.Open("MyLog.txt", FileMode.Open, FileAccess.Read, FileShare.None);```

> LogManager.Configuration = null;

It works for one test but if I have parameterized test [TheoryData], it does work for all tests.

If it helps, I wrote the different use cases:
```

using NLog;
using NLog.Config;
using NLog.Targets;
using System.IO;
using System.Text;
using System.Threading.Tasks;
using Xunit;

public static class MyLogExtension
{
    static MyLogExtension()
    {
        var config = new LoggingConfiguration();
        string myLog = "MyLog";

        var myTarget = new FileTarget(myLog);
        myTarget.FileName = "MyLog.txt";
        myTarget.Encoding = Encoding.UTF8;
        myTarget.NetworkWrites = false;
        myTarget.ConcurrentWrites = true;
        myTarget.KeepFileOpen = true;
        myTarget.Header = "#####################################${newline}LOG OPENED [${longdate}]${newline}";
        myTarget.Footer = "LOG CLOSED [${longdate}]${newline}=====================================${newline}";
        myTarget.Layout = "${message}${onexception:inner=${newline}";

        config.AddTarget(myTarget);

        var ruleForInternal = new LoggingRule(myLog, LogLevel.Debug, myTarget);

        config.LoggingRules.Add(ruleForInternal);

        LogManager.Configuration = config;
    }

    public static void Instanciate()
    {

    }

    public static void Log(string messsage)
    {
        NLog.Logger logger = LogManager.GetLogger("MyLog");
        logger.Info(messsage);
    }
}

public class MyClass
{
    public void WriteSomething(string text)
    {
        MyLogExtension.Log($"Error: {text}");
    }
}

public class MyClassTest
{
    [Theory]
    [InlineData("whatever", "Error: whatever")]
    [InlineData("whatever 2", "Error: whatever 2")]
    [InlineData("whatever 3", "Error: whatever 3")]
    public void SimpleTest(string message, string expectedErrorMessage)
    {
        // Arrange
        MyClass classUnderTest = new MyClass();

        // Act
        Task.WaitAll( // Needed?
            Task.Factory.StartNew(() => { classUnderTest.WriteSomething(message); })
        );
        
        Assert.Equal(true, File.Exists("MyLog.txt"));
        var stream = File.Open("MyLog.txt", FileMode.Open, FileAccess.Read, FileShare.None);
        //Result Message:	System.IO.IOException : The process cannot access the file 'MyLog.txt' because it is being used by another process.
    }


    [Theory]
    [InlineData("whatever", "Error: whatever")]
    [InlineData("whatever 2", "Error: whatever 2")]
    [InlineData("whatever 3", "Error: whatever 3")]
    public void SimpleTest_reset_KeepFileOpen(string message, string expectedErrorMessage)
    {
        MyLogExtension.Instanciate();

        FileTarget target = LogManager.Configuration.FindTargetByName("MyLog") as FileTarget;
        target.KeepFileOpen = false;
        LogManager.ReconfigExistingLoggers();

        // Arrange
        MyClass classUnderTest = new MyClass();

        // Act
        Task.WaitAll( // Needed?
            Task.Factory.StartNew(() => { classUnderTest.WriteSomething(message); })
        );
        
        Assert.Equal(true, File.Exists("MyLog.txt"));
        var stream = File.Open("MyLog.txt", FileMode.Open, FileAccess.Read, FileShare.None);
        //Result Message:	System.IO.IOException : The process cannot access the file 'MyLog.txt' because it is being used by another process.
    }

    [Theory]
    [InlineData("whatever", "Error: whatever")]
    [InlineData("whatever 2", "Error: whatever 2")]
    [InlineData("whatever 3", "Error: whatever 3")]
    public void SimpleTest_reset_Nlog(string message, string expectedErrorMessage)
    {
        // Arrange
        MyClass classUnderTest = new MyClass();

        // Act
        Task.WaitAll( // Needed?
            Task.Factory.StartNew(() => { classUnderTest.WriteSomething(message); })
        );

        LogManager.Configuration = null;

        Assert.Equal(true, File.Exists("MyLog.txt"));
        var stream = File.Open("MyLog.txt", FileMode.Open, FileAccess.Read, FileShare.None);
    }
}
```

**Comment by 304NotModified** [comment]
> It works for one test but if I have parameterized test [TheoryData], it does work for all tests.

Do you get an error? What isn't working?

**Comment by 22u2** [comment]
> Do you get an error? What isn't working?

I have the error message saying that a process is locking the log file.

**Comment by 304NotModified** [comment]
Thanks for tests. Testing now

**Comment by 304NotModified** [comment]
issue the issue. We are closing the configuration (by `LogManager.Configuration = null;`), and we never reinitialize the config. So the second `LogManager.Configuration = null;` (or `LogManager.Configuration.Close()` will not close the fileTarget.

Have to dig in some deeper

**Comment by 304NotModified** [comment]
LOL see it. It's not NLog, but you are keeping the stream opened ;)

```
var stream = File.Open("MyLog.txt", FileMode.Open, FileAccess.Read, FileShare.None);
            stream.Close(); //<! ----------------------------------- need this
```

**Comment by 304NotModified** [comment]
(mmm questions are never invalid)

TL;DR, not a bug in NLog, but it the test  (IMO) :)

**Comment by 22u2** [comment]
Oh I thought I was using File.ReadAllText
Thx for your time!
